openssl-dynamic/src/main/c/io_netty_tcnative_jni_ocsp_Native.c

```diff
@@ -0,0 +1,314 @@
+  if (CALLBACK_CLAZZ != NULL) {
+    (*env)->DeleteGlobalRef(env, CALLBACK_CLAZZ);
+
+    CALLBACK_CLAZZ = NULL;
+    CALLBACK_METHOD = NULL;
+  }
+}
+
+/**
+ * Installs the OCSP stapling callback.
+ */
+JNIEXPORT jobject JNICALL Java_io_netty_tcnative_jni_ocsp_Native_newCallback
+    (JNIEnv *env, jclass clazz, jlong ctx, jboolean client, jobject callback) {
+
+  long error = -1L;
+  
+  if (ctx == 0L) {
+    fprintf(stderr, "ctx ptr is 0L");
+    return NULL;
+  }
+
+  //
+  // ATTENTION: We need to create a global reference of the callback.
+  //
+  jweak callbackRef = (*env)->NewWeakGlobalRef(env, callback);
+
+  tcn_ssl_ctxt_t *c = J2P(ctx, tcn_ssl_ctxt_t *);
+  
+  error = SSL_CTX_set_tlsext_status_arg(c->ctx, callbackRef);
+  if (error <= 0L) {
+    fprintf(stderr, "newCallback/SSL_CTX_set_tlsext_status_arg: %li\n", error);
+    freeCallback(env, NULL, callbackRef);
+    return NULL;
+  }
+  
+  //
+  // The client and server use slightly different return values.
+  // In particular in the event of an error. We need therefore
+  // two slightly different callback functions.
+  //
+  
+  if (client == JNI_TRUE) {
+    error = SSL_CTX_set_tlsext_status_cb(c->ctx, io_netty_tcnative_jni_ocsp_client_callback);
+    if (error <= 0L) {
+      fprintf(stderr, "newCallback/SSL_CTX_set_tlsext_status_cb (client): %li\n", error);
+      freeCallback(env, NULL, callbackRef);
+      return NULL;
+    }
+  
+  } else {
+    error = SSL_CTX_set_tlsext_status_cb(c->ctx, io_netty_tcnative_jni_ocsp_server_callback);
+    if (error <= 0L) {
+      fprintf(stderr, "newCallback/SSL_CTX_set_tlsext_status_cb (server): %li\n", error);
+      freeCallback(env, NULL, callbackRef);
+      return NULL;
+    }
+  }
+
+  return callbackRef;
+}
+
+/**
+ * Releases the OCSP callback.
+ */
+JNIEXPORT void JNICALL Java_io_netty_tcnative_jni_ocsp_Native_freeCallback
+    (JNIEnv *env, jclass clazz, jlong ctx, jobject callback) {
+
+  tcn_ssl_ctxt_t *c = J2P(ctx, tcn_ssl_ctxt_t *);
+  freeCallback(env, c, callback);
+}
+
+/**
+ * Enables or disables OCSP stapling for the given SSLEngine.
+ */
+JNIEXPORT jboolean JNICALL Java_io_netty_tcnative_jni_ocsp_Native_setType
+    (JNIEnv *env, jclass clazz, jlong ssl, jint type) {
+
+  if (ssl == 0L) {
+    fprintf(stderr, "SSL ptr is 0L");
+    return JNI_FALSE;
+  }
+
+  long error = SSL_set_tlsext_status_type((SSL*)ssl, type);
+  if (error <= 0L) {
+    fprintf(stderr, "setType/SSL_set_tlsext_status_type: %li\n", error);
+  }
+  
+  return (error == 1L) ? JNI_TRUE : JNI_FALSE;
+}
+
+/**
+ * Server: Sets the stapled OCSP record bytes for the given SSLEngine.
+ */
+JNIEXPORT jboolean JNICALL Java_io_netty_tcnative_jni_ocsp_Native_setResponse
+    (JNIEnv *env, jclass clazz, jlong ssl, jbyteArray response) {
+  
+  if (ssl == 0L) {
+    fprintf(stderr, "SSL ptr is 0L");
+    return JNI_FALSE;
+  }
+
+  jsize length = (*env)->GetArrayLength(env, response);
+  if ((*env)->ExceptionCheck(env)) {
```

Comment:
can you explain why this is necessary? If this is necessary I'm guessing we will need to do this more globally.